### ISSUE TITLE
examples /08-combined-potential.py doesn't return the same energy as in examples /coulomb_test_frames.xyz

### ISSUE THREAD

**Issue by YuchiGuo** [lab-cosmo/torch-pme#185]
examples /08-combined-potential.py doesn't return the same energy as in examples /coulomb_test_frames.xyz. 

For example, for the second dataset in coulomb_test_frames.xyz, the energy is stated to be -19.86503067171332. But when I run compute_energy() in 08-combined-potential.py, the energy is -26.7268. 

Is this discrepancy intended?

**Comment by YuchiGuo** [comment]
In addition, when I used the code in examples/01-charges-example.py to calculate energy for the positions and charges given in coulomb_test_frames.xyz, the resultant energy is 10-100 times smaller than the energy stated in the xyz file.

**Comment by E-Rum** [comment]
@YuchiGuo thank you for your comments! The thing is, Ewald summation (both the classical version and the mesh-based one) is highly sensitive to hyperparameters. The example scripts are there to demonstrate a specific use case, not necessarily with ideal parameters. If you're interested in matching theoretical values, could you please explicitly provide the parameters you're using? Then we can try to find the actual optimal parameters together.

**Comment by E-Rum** [comment]
Can you maybe try decreasing mesh_spacing and see if the energy gets closer to the desired value?

**Comment by E-Rum** [comment]
The example with 08-combined-potential.py is fixed now. Indeed, lr_wavelength was not small enough for convergence to the theoretical values—thanks for the catch!

**Comment by YuchiGuo** [comment]
Thanks but 08-combined-potential.py still returns -26 instead of -19?

**Comment by E-Rum** [comment]
Can you clarify exactly which line you're talking about? The thing is, if you're using a combined potential with arbitrary weights in front of the individual potentials, then you won't yield Coulomb-like values. If you just initialize your combined potential with exponents 1 and 2, it will simply be a weighted combination of these exponents.

**Comment by YuchiGuo** [comment]
Ah this makes sense! I didn't quite realise that the combined potential is different to Coulomb potential. So this problem is sorted.

Lastly, could you clarify what is the origin of the energy of -19 for the second dataset in coulomb_test_frames.xyz -- is it not purely coulomb? Because when I used PMECalculator to calculate energy, I got a value of roughly 1.

**Comment by E-Rum** [comment]
It's purely Coulomb. Probably, as I said before, the PME parameters might not be tight enough. Try decreasing mesh_spacing, and let me know if it helps :)

**Comment by YuchiGuo** [comment]
Hi! i tried decreasing mesh_spacing and the energy still hovers around 1.

**Comment by YuchiGuo** [comment]
Thank you for your help! Please see the code below. It's almost the same as examples/01-charges-example.py except where I manually inputted positions, charges, and parameters like smearing and mesh_spacing. 


import torch
import vesin.torch
#import vesin.torch.metatensor
#from metatensor.torch import Labels, TensorBlock, TensorMap
#from metatensor.torch.atomistic import NeighborListOptions, System

import torchpme
from torchpme.tuning import tune_pme

dtype = torch.float64

# %%
#
# Create the properties CsCl unit cell
symbols = ("Cs", "Cl")
types = torch.tensor([55, 17])
#charges = torch.tensor([[1.0], [-1.0]], dtype=dtype)
#positions = torch.tensor([(0, 0, 0), (0.5, 0.5, 0.5)], dtype=dtype)
cell = 8*torch.eye(3, dtype=dtype)
pbc = torch.tensor([True, True, True])


charges = torch.tensor([[1.0],[1.0], [1.0],[1.0],[-1.0],[-1.0],[-1.0],[-1.0]], dtype=dtype)
 
positions = torch.tensor( [(7.50000000,       1.50999999,       0.05000000),
                           (6.37000036  ,     6.44000006  ,     5.28999996),    
                           (1.15999997   ,    2.56000018   ,    3.88999987),    
                           (5.46000004    ,   7.88000011    ,   6.57999992),
                           (3.61999989     ,  5.91000032     ,  3.55999994),
                           (6.17000008      , 4.02999973  ,     2.26999998),
                           (4.38000011       ,0.11000000   ,    0.82000005),
                           (0.71000004       ,2.49000001    ,   0.93999994) ], dtype=dtype)

# %%
#
# Based on our system we will first *tune* the PME parameters for an accurate computation.
# The ``sum_squared_charges`` is equal to ``2.0`` becaue each atom either has a charge
# of 1 or -1 in units of elementary charges.

cutoff = 4
nl = vesin.torch.NeighborList(cutoff=cutoff, full_list=False)
neighbor_indices, neighbor_distances = nl.compute(
    points=positions.to(dtype=torch.float64, device="cpu"),
    box=cell.to(dtype=torch.float64, device="cpu"),
    periodic=True,
    quantities="Pd",
)

print(neighbor_indices[1].item)
#for i in range(len(neighbor_indices)):
 #   print(neighbor_indices[i].data)

#print("neighbor_indices", neighbor_indices)
#print("neighbor_distances", neighbor_distances.size())
smearing, pme_params, _ = tune_pme(
    charges=charges,
    cell=cell,
    positions=positions,
    cutoff=cutoff,
    neighbor_indices=neighbor_indices,
    neighbor_distances=neighbor_distances,
)

# %%
#
# The tuning found the following best values for our system.

print("smearing:", smearing)
print("PME parameters:", pme_params)
print("cutoff:", cutoff)

# %%
#
# Based on the system we compute the corresponding half neighbor list using `vesin
# <https://luthaf.fr/vesin>`_ and rearrange the results to be suitable for the
# calculations below.

nl = vesin.torch.NeighborList(cutoff=cutoff, full_list=False)

neighbor_indices, S, D, neighbor_distances = nl.compute(
    points=positions, box=cell, periodic=True, quantities="PSDd"
)

print(neighbor_distances)
print(neighbor_indices)

# %%
#
# Next, we initialize the :class:`PMECalculator` calculator with an ``exponent`` of
# *1* for electrostatic interactions between the two atoms. This calculator
# will be used to *compute* the potential energy of the system.

calculator = torchpme.PMECalculator(
    torchpme.CoulombPotential(smearing=0.66666667), interpolation_nodes= 4, mesh_spacing=1.666666666666667 #**pme_params
)
calculator.to(dtype=dtype)

# %%
#
# Single Charge Channel
# #####################
#
# As a first application of multiple charge channels, we start simply by using the
# classic definition of one charge channel per atom.

#charges = torch.tensor([[1.0], [-1.0]], dtype=dtype)

# %%
#
# Any input the calculators has to be a 2D array where the *rows* describe the number of
# atoms (here ``(2)``) and the *columns* the number of atomic charge channels (here
# ``(1)``).

print(charges.shape)

# %%
#
# Calculate the potential using the PMECalculator calculator

potential = calculator(
    positions=positions,
    cell=cell,
    charges=charges,
    neighbor_indices=neighbor_indices,
    neighbor_distances=neighbor_distances,
)

k_potential = calculator._compute_kspace(charges,cell, positions)
print("k_potential is {}".format(k_potential)) 

# %%
#
# We find a potential that is close to the Madelung constant of a CsCl crystal which is
# :math:`2 \cdot 1.76267 / \sqrt{3} \approx 2.0354`.

print("r_potential is {}".format(potential-k_potential)) 
print("Total potential per atom is {}".format(potential))

print("Total potential is {}".format(charges.T @ potential)) #output is tensor([[-1.4349]], dtype=torch.float64)

**Comment by YuchiGuo** [comment]
Thank you for your help! Please see the code below. It's almost the same as examples/01-charges-example.py except where I manually inputted positions, charges, and parameters like smearing and mesh_spacing.

[01-charges-example copy.txt](https://github.com/user-attachments/files/20040085/01-charges-example.copy.txt)

**Comment by E-Rum** [comment]
Thank you! The problem is that in your code, to match the actual value, we also need to include the prefactor in the calculator:

prefactor = torchpme.prefactors.eV_A

I hope this helps

**Comment by YuchiGuo** [comment]
This indeed solves the problem! Thanks a lot!
